File: 238_product_array_except_self/product_array_except_self.py
from typing import List

class Solution:
    def product_except_self(self, nums: List[int]) -> List[int]:
        n = len(nums)
        answer = [1] * n

        for i in range(1, n):
            answer[i] = answer[i - 1] * nums[i - 1]

        right_product = 1
        for i in range(n - 1, -1, -1):
            answer[i] *= right_product
            right_product *= nums[i]

        return answer

if __name__ == "__main__":
    solution = Solution()
    nums = [1,2,3,4]
    result = solution.product_except_self(nums)
    print(result)

File: 238_product_array_except_self/product_array_except_self.go
package main

import "fmt"

type Solution struct {
}

func (*Solution) productExceptSelf(nums []int) []int {
	n := len(nums)
	answer := make([]int, n)

	answer[0] = 1
	for i := 1; i < n; i++ {
		answer[i] = answer[i-1] * nums[i-1]
	}

	rightProduct := 1
	for i := n - 1; i >= 0; i-- {
		answer[i] *= rightProduct
		rightProduct *= nums[i]
	}

	return answer
}

func main() {
	solution := new(Solution)
	nums := []int{1, 2, 3, 4}
	result := solution.productExceptSelf(nums)
	fmt.Println(result)
}
